book/src/proposals/leader-duplicate-block-slashing.md

```diff
@@ -0,0 +1,219 @@
+
+An issue to note here is that if some version of a slot violates correctness
+before the slot is finished, then the validator does not know what the ending
+blockhash is and thus cannot store the slot as dead. In these cases we can drop
+the entire slot and wait for repair. More details in the `Replay Failures` 
+section below.
+
+### Repairing Multiple versions of the Same Slot
+Repair is augmented with a blockhash. The various types of repairs:
+
+`Shred(slot, blockhash, index)` - Ask for a specific shred in a slot
+with version `blockhash`. 
+
+Notes:
+
+1) Setting `blockhash` to `Hash::default()` means this validator assumes there
+is only one version of this slot in the cluster and will acccept any shred
+for `slot` and `index`. However, if a validator gets a mixed bag of such
+shreds that do not chain, they will then drop the entire `slot`, and only
+repair `slot` if some later child slot chains back to `slot` (this will be 
+initiated by `Orphan` repairs below). 
+
+2) Repair responses need to be tied to a particular blockhash. However, 
+because repair responses are limited to the size of a shred, we cannot
+include the blockhash in the repair response. This means we need to include 
+some repair "cookie" in the request + response that maps some number of bits
+to a particular `blockhash`. This breaks down if there are more versions of 
+this slot than can be tracked by the number of bits allocated to the cookie.
+
+
+`Orphan(slot, first_tick_hash, num_hashes)` - Ask for the last shred
+of the parent of the child slot, where the child has slot equal to `slot`. 
+This last shred must contain the last tick `T_p` in that parent slot such that
+hashing `T_p` `num_hashes` times is equal to `first_tick_hash`.
+
+Notes:
+
+1) The response will include the last shred of the parent slot. From this
+last shred the validator should be able to get the last tick and the 
+merkle root, which can be used to compute the `blockhash`. This `blockhash`
+is then used to make requests of the form `Shred(slot, blockhash, index)`
+to repair the rest of this slot.
+
+2) Also requires a cookie similar to the `Shred(slot, blockhash, index)` 
+requests above.
+
+### Chaining Multiple Versions of the Same Slot
+When there are multiple versions for a slot `A`, a natural question that arises
+is how validators know which version of slot `A` some descendant slot `B`
+chains to. Shreds currently specify only a `parent` slot, but not which version 
+of that parent. The approach to figure out chaining behavior is then:
+
+1) Wait for all the shreds for the first entry `E_B` of slot `B` to arrive 
+(Implementation can make sure first shred `S_B` always contains only a tick 
+to avoid waiting for multiple shreds).
+2) These shreds for slot `B` are stored under the version `Hash::default()`
+(optimistically assume this child is the only version). If another conflicting
+version of `B` is detected before this version is completed, we drop all the
+shreds for slot `B`.
```

Comment:
Need to reason through correctness here. If we have multiple versions of a slot `C` chaining to multiple version of a slot `B`, chaining to multiple versions of a slot `A`, will the procedure we have outlined thus far handle it?

Reply:
should do, because we can only chain forward one entry at a time

Reply:
The tricky part is whether the below is correct:
1) Dropping any unfinished slots on detecting another version
2) Waiting on repair to perform chaining on slots with multiple versions based on blockhash

I think it should work as well :D